pkg/agent/ipassigner/ip_assigner_linux.go

```diff
@@ -211,99 +387,156 @@ func (a *ipAssigner) UnassignIP(ip string) (bool, error) {
+		if err := as.destroy(); err != nil {
+			return err
 		}
+		delete(a.vlanAssignees, subnetInfo.VLAN)
 	}
-
-	a.assignedIPs.Delete(ip)
-	return true, nil
+	delete(a.assignedIPs, ip.String())
+	return nil
 }
 
 // AssignedIPs return the IPs that are assigned to the dummy device.
-func (a *ipAssigner) AssignedIPs() sets.Set[string] {
+func (a *ipAssigner) AssignedIPs() map[string]*crdv1b1.SubnetInfo {
 	a.mutex.RLock()
 	defer a.mutex.RUnlock()
 	// Return a copy.
-	return a.assignedIPs.Union(nil)
+	copy := map[string]*crdv1b1.SubnetInfo{}
+	for k, v := range a.assignedIPs {
+		copy[k] = v
+	}
+	return copy
 }
 
-// InitIPs loads the IPs from the dummy device and replaces the IPs that are assigned to it
+// InitIPs loads the IPs from the dummy/vlan devices and replaces the IPs that are assigned to it
 // with the given ones. This function also adds the given IPs to the ARP/NDP responder if
 // applicable. It can be used to recover the IP assigner to the desired state after Agent restarts.
-func (a *ipAssigner) InitIPs(ips sets.Set[string]) error {
-	a.mutex.Lock()
-	defer a.mutex.Unlock()
-	if a.dummyDevice != nil {
-		assigned, err := a.loadIPAddresses()
-		if err != nil {
-			return fmt.Errorf("error when loading IP addresses from the system: %v", err)
-		}
-		for ip := range ips.Difference(assigned) {
-			addr := util.NewIPNet(net.ParseIP(ip))
-			if err := netlink.AddrAdd(a.dummyDevice, &netlink.Addr{IPNet: addr}); err != nil {
-				if !errors.Is(err, unix.EEXIST) {
-					return fmt.Errorf("failed to add IP %v to interface %s: %v", ip, a.dummyDevice.Attrs().Name, err)
-				}
-			}
-		}
-		for ip := range assigned.Difference(ips) {
-			addr := util.NewIPNet(net.ParseIP(ip))
-			if err := netlink.AddrDel(a.dummyDevice, &netlink.Addr{IPNet: addr}); err != nil {
-				if !errors.Is(err, unix.EADDRNOTAVAIL) {
-					return fmt.Errorf("failed to delete IP %v from interface %s: %v", ip, a.dummyDevice.Attrs().Name, err)
-				}
-			}
-		}
+func (a *ipAssigner) InitIPs(desired map[string]*crdv1b1.SubnetInfo) error {
+	if err := a.loadIPAddresses(); err != nil {
+		return fmt.Errorf("error when loading IP addresses from the system: %v", err)
 	}
-	for ipStr := range ips {
-		ip := net.ParseIP(ipStr)
-		var err error
-		if utilnet.IsIPv4(ip) && a.arpResponder != nil {
-			err = a.arpResponder.AddIP(ip)
-		}
-		if utilnet.IsIPv6(ip) && a.ndpResponder != nil {
-			err = a.ndpResponder.AddIP(ip)
+	staleIPs := sets.StringKeySet(a.assignedIPs)
+	for ip, desiredSubnetInfo := range desired {
+		if _, err := a.AssignIP(ip, desiredSubnetInfo, true); err != nil {
+			return err
 		}
-		if err != nil {
+		staleIPs.Delete(ip)
+	}
+	for ip := range staleIPs {
+		if _, err := a.UnassignIP(ip); err != nil {
 			return err
 		}
-		a.advertise(ip)
 	}
-	a.assignedIPs = ips.Union(nil)
 	return nil
 }
 
+func (a *ipAssigner) GetInterfaceID(subnetInfo *crdv1b1.SubnetInfo) (int, bool) {
+	as, _ := a.getAssignee(subnetInfo, false)
```

Comment:
not protected by read lock?

Reply:
my bad, missed it when adding the method